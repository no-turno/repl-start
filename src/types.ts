import "../types/src/types.d.ts"