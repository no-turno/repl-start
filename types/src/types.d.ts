/**
 * Creates a start configuration object with a generic type T.
 * @example
 * ```typescript
 * const startConfig = createStartConfig({ name: 'John', age: 30 });
 * console.log(startConfig.config); // { name: 'John', age: 30 }
 * ```
 */
export declare function createStartConfig<T = Record<string, any>>(config: T): {
    config: T;
};
/**
 * Parses the input configuration and returns the parsed result.
 *
 * @example
 * ```typescript
 * const startConfig = createStartConfig({ name: 'John', age: 30 });
 * const parsedConfig = parseConfig(startConfig);
 * console.log(parsedConfig); // { name: 'John', age: 30 }
 * ```
 */
export declare function parseConfig(config: ReturnType<typeof createStartConfig> & {
    raw: string;
}): any;
/**
 * Reads a configuration value from the input configuration object using a query string.
 *
 * @example
 * ```typescript
 * const startConfig = createStartConfig({ name: 'John', age: 30 });
 * const parsedValue = readConfig(startConfig, '.name');
 * console.log(parsedValue); // 'John'
 * ```
 */
export declare function readConfig(config: any, query: string): string;
/**
 * Constructs a query string from a root string and an array of path segments.
 *
 * @example
 * ```typescript
 * const root = `{ 'name': "john" }`;
 * const path = ['name'];
 * const query = $q(root, ...path) // any;
 * console.log(query); // any for `(({ 'name': "john" })['name'])` \/* john *\/
 * ```
 */
declare function $q(root: TemplateStringsArray, ...path: string[]): string;
declare const _default: {
    createStartConfig: typeof createStartConfig;
    parseConfig: typeof parseConfig;
    readConfig: typeof readConfig;
    $q: typeof $q;
};
export default _default;
