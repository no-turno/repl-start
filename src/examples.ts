import { $q, createStartConfig, parseConfig, readConfig } from ".";

const startConfig = createStartConfig({
    bundler: {
        bun: {
            config: {
                transpiler: {
                    jsx: "react",
                    jsxFactory: "React.createElement",
                    jsxFragment: "React.Fragment",
                },
            },
        },
    },
});
console.log(startConfig);

const config = parseConfig({ ...startConfig, raw: Bun.inspect(startConfig) });
console.log(config);

// reading from some config
const bundler = readConfig(config, $q`config.bundler`)
console.log(bundler);


// reading from previous querying in config
const bun = readConfig(bundler, $q`bun.config`)
console.log(bun);
