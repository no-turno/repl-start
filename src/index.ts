/**
 * Creates a start configuration object with a generic type T.
 * @example
 * ```typescript
 * const startConfig = createStartConfig({ name: 'John', age: 30 });
 * console.log(startConfig.config); // { name: 'John', age: 30 }
 * ```
 */
export function createStartConfig<T = Record<string, any>>(config: T):T{
    return config
}


/**
 * Parses the input configuration and returns the parsed result.
 *
 * @example
 * ```typescript
 * const startConfig = createStartConfig({ name: 'John', age: 30 });
 * const parsedConfig = parseConfig(startConfig);
 * console.log(parsedConfig); // { name: 'John', age: 30 }
 * ```
 */
export function parseConfig(config: ReturnType<typeof createStartConfig> | { 
    raw: string }): any {
    return new ShadowRealm().evaluate(`const out = \`${Bun.inspect(config)}\`;out`)
}


/**
 * Reads a configuration value from the input configuration object using a query string.
 *
 * @example
 * ```typescript
 * const startConfig = createStartConfig({ name: 'John', age: 30 });
 * const parsedValue = readConfig(startConfig, '.name');
 * console.log(parsedValue); // 'John'
 * ```
 */
export function readConfig(config: any, query: string): string {
    return new ShadowRealm().evaluate(`Bun.inspect((${config})${query})`)
}

/**
 * Constructs a query string from a root string and an array of path segments.
 *
 * @example
 * ```typescript
 * const root = `{ 'name': "john" }`;
 * const path = ['name'];
 * const query = $q(root, ...path) // any;
 * console.log(query); // any for `(({ 'name': "john" })['name'])` \/* john *\/
 * ```
 */
export function $q(root: TemplateStringsArray, ...path: string[]): string {
    return (root.toString().concat(path.join())).split('.').map(k => Bun.inspect([k])).join('')
}

export default {
    createStartConfig,
    parseConfig,
    readConfig,
    $q
}