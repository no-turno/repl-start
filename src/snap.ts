export default `{
        {
            let _bundler = {
        bun: {
            config: {
                transpiler: {
                    jsx: "react",
                    jsxFactory: "React.createElement",
                    jsxFragment: "React.Fragment",
                },
            },
        }
    };

    console.log(_bundler);

    let _bun = {
        transpiler: {
            jsx: "react",
            jsxFactory: "React.createElement",
            jsxFragment: "React.Fragment",
        },
    };

    console.log(_bun)
}
}`
