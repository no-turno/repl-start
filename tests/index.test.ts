import {test, it,describe, expect} from "bun:test"
import { $q, createStartConfig, parseConfig, readConfig } from "../src"

describe('config methods', () => {
    test("should return config object", () => {
        const config = createStartConfig({
            test: true
        })
        expect(config).toMatchObject({
            test: true
        })
    })

    test("should parse config object", () => {
        const config = parseConfig(createStartConfig({
            test: true
        }))

        expect(config).toMatchSnapshot(Bun.inspect({
            test: true
        }))
    })

    test("should query parsed config object", () => {
        const config = parseConfig(createStartConfig({
            test: true
        }))
        expect(readConfig(config, $q`test`) === "true").toBeTrue()
    })
})