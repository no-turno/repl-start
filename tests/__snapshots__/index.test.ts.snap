// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`config methods should parse config object: {
  test: true,
} 1`] = `
"{
  test: true,
}"
`;
